DX-1264: Update Exports & Add CI Tests

ytkimirti opened the pull request:
This PR has three main changes:
- update eslint & exports in package.json
- fixing small errors in examples and adding extensive CI
- llamaindex is removed, as it was causing errors explained below

Since llamaindex is removed, we will release a major version (2.0.0)

## Eslint & Exports

- Updated tsup config
- Updated exports field in package.json
>NOTE: This should not bring any breaking changes with the old version. Here is the old and new output coming from the attw cli:
<img width="625" alt="rag before" src="https://github.com/user-attachments/assets/0e5e0569-b496-498f-895f-2b4ee5a5878d">
<img width="620" alt="rag after" src="https://github.com/user-attachments/assets/fc526775-553c-4cd9-a1ec-57c33d46fe32">

## CI

The examples are first tested with a local build. Then, a release is made with ci tag and it's released. It's tested by deploying to Vercel and CF workers.

fahreddinozcan committed:
update eslint

ytkimirti committed:
fix: update eslint plugin versions and fix rules

ytkimirti committed:
fix: turn optional deps into normal deps

ytkimirti committed:
fix: husky install warning

ytkimirti committed:
fix: rsc exports and tsup config

ytkimirti committed:
chore: format eslint

ytkimirti committed:
fix: ignore examples dir in eslint

ytkimirti committed:
chore: check examples

ytkimirti committed:
fix: use wider versions in peer deps

ytkimirti committed:
chore: use lint-staged because of eslint error

ytkimirti committed:
chore: lockfile

ytkimirti committed:
chore: add comment on rsc import

ytkimirti committed:
chore: remove unused comment

ytkimirti committed:
fix: remove Prettify type utility because of problems with eslint

ytkimirti committed:
chore: remove unused flag

it was only for pnpm, there is no such flag in npm

ytkimirti committed:
fix: remove types field in exports

ytkimirti committed:
chore: lockfile

CahidArda commented on examples/cloudflare-workers/README.md:
````diff
@@ -28,13 +28,7 @@ npm install @upstash/rag-chat
-In older CF worker versions, you may need to set the following compatibility flags:
-
-```toml
-compatibility_flags = [ "streams_enable_constructors", "transformstream_enable_standard_constructor" ]
````
why do we delete these? I needed them for streaming, did you try streaming with only the nodejs_compat_v2 flag?

ytkimirti replied:
Yes it works, why did you need them, shouldn't they be default?
https://developers.cloudflare.com/workers/configuration/compatibility-dates/#streams-constructors
https://developers.cloudflare.com/workers/configuration/compatibility-dates/#compliant-transformstream-constructor

CahidArda replied:
`In older CF worker versions` :D

let's delete them then, if someone has any issues, we can let them know about the flags.

ytkimirti replied:
Ohh makes sense, I will add it back

ytkimirti committed:
chore: fix homepage url

ytkimirti committed:
fix: add missing vector dep to cloudflare example

ytkimirti committed:
fix: add cloudflare example openai as comment instead

CahidArda commented on examples/cloudflare-workers/src/index.ts:
```diff
@@ -1,8 +1,117 @@
+
+app.use("*", async (c, next) => {
+  const ragChat = new RAGChat({
+    // model: upstash("meta-llama/Meta-Llama-3-8B-Instruct", { apiKey: c.env.QSTASH_TOKEN }),
```
we can put the model back in, and comment out the alternative below

ytkimirti replied:
Oh i missed this thanks

ytkimirti commented on package.json:
```diff
@@ -71,19 +77,17 @@
     "pdf-parse": "^1.1.1",
-    "unstructured-client": "^0.15.1"
+    "unstructured-client": "^0.15.1",
+    "@langchain/anthropic": "^0.2.15",
+    "@langchain/mistralai": "^0.0.28",
+    "langsmith": "^0.1.41"
```
tsup does not bundle dependencies by default so we don't have to label them as external. these will just be normal dependencies.

We can turn them into optional peer dependencies in the future, that's probably the best way to do it but we need dynamic imports for that. And ofc that would be a breaking change so in the future

ytkimirti commented on tsup.config.ts:
```diff
@@ -4,32 +4,21 @@ export default defineConfig([
   },
   {
-    entry: ["src/nextjs/rsc-server.ts", "src/nextjs/rsc-client.ts"],
+    entry: ["src/nextjs/rsc-server.ts"],
```
yepp, actually there is no special key for rsc-client. Only special import is for "react-server" so we actually can't make react use "rsc-client.ts" and normal imports to use "index.ts"

ytkimirti commented on src/nextjs/index.ts:
```diff
@@ -1,3 +1,7 @@
+export * from "./chat-adapter";
+export * from "./server-action-read-adapter";
+
+// This cant be exported here because you can't import ai/rsc
+// when not using a server component
+// Opposite is true for the read adapter, it can't be imported in the rsc-server.ts file
```
Yeah, I should make a better explanation in the commet

ytkimirti committed:
chore: improve comment for rsc-server.ts

ytkimirti committed:
chore: add back cf compat info to readme

ytkimirti committed:
fix: cf example

ytkimirti committed:
feat: add video link to the landing page of chat-to-website example

ytkimirti committed:
fix: small fix in example

ytkimirti committed:
fix: add node10 support for imports

ytkimirti committed:
fix: revert

ytkimirti committed:
fix: cli

ytkimirti committed:
chore: update version for chat-to-website example

ytkimirti committed:
chore: remove lockfiles for chat-to-website

CahidArda commented:
[more on react-server export](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md)

CahidArda commented:
✘ [ERROR] The `nodejs_compat_v2` compatibility flag is experimental and must be prefixed with `experimental:`. Use `experimental:nodejs_compat_v2` flag instead.

ytkimirti committed:
fix: update cf example compat date

ytkimirti committed:
chore: add note about the bundle size on cf example

ytkimirti committed:
chore: add cf test to ci

ytkimirti committed:
fix: ci

ytkimirti committed:
fix: ci only test src file

ytkimirti committed:
chore: install @types/bun for example ci file warnings

ytkimirti committed:
feat: add nodejs ci test

ytkimirti committed:
fix: ci

ytkimirti committed:
fix: ci

ytkimirti committed:
fix: flaky tests

ytkimirti committed:
chore: small update for cf example

ytkimirti committed:
fix: ci

ytkimirti committed:
fix: ci

ytkimirti committed:
fix: ci

CahidArda committed:
fix: use latest rag-chat in nextjs examples

CahidArda committed:
fix: add local tests for nextjs examples

CahidArda committed:
fix: add DEPLOYMENT_URL to tests

CahidArda committed:
feat: add prerelease step to tests ci

CahidArda committed:
fix: localhost ports in ci

ytkimirti committed:
fix: bump llamaindex

CahidArda committed:
feat: add deploy test for vercel

CahidArda committed:
fix: replace localhost

CahidArda committed:
Revert "fix: bump llamaindex"

This reverts commit 5ab17c3fd664114f7e2abbfd574a678f899ea560.

CahidArda committed:
fix: add working dir

CahidArda commented:
managed to reproduce the `Error: Missing tiktoken_bg.wasm` error (mentioned in https://github.com/upstash/rag-chat/issues/82) in the CI test of [20e3c11](https://github.com/upstash/rag-chat/pull/79/commits/20e3c11f4f1a282cf0e005eacd96387de78ae444).

[See](https://github.com/upstash/rag-chat/actions/runs/11147944037/job/30991074857)

CahidArda committed:
fix: pin llamaindex to 0.5.24

CahidArda commented:
Some notes on the error:
- the error we get depends on the package manager used in the app (npm vs pnpm)

| - | - | - | - |
| rag-chat version | llamaindex version | Vercel (npm) | Vercel (pnpm) |
| 1.6.0 | ^0.5.20 | ❌ llamaindex/openai version missing | I didn't try. Probably the same error as the one to the left |
| 1.6.1 | ^0.6.0 | ✅ | ❌ unexpected char |
| 1.6.4 | 0.5.24 | ✅ | ❌ [Missing tiktoken_bg.wasm](https://github.com/run-llama/create-llama/issues/164) |

the missing file has a task at Vercel https://github.com/vercel/next.js/issues/54395 and at tiktoken https://github.com/dqbd/tiktoken/issues/67.

The unexpected char problem:
- referenced in @xenova/transformers https://github.com/xenova/transformers.js/issues/210 and they suggest updating dependencies. I tried both dependencies and peerDependencies. It didn't work.
- potential solutions in https://github.com/xenova/transformers.js/issues/338. serverComponentsExternalPackages ([1](https://github.com/xenova/transformers.js/issues/210#issuecomment-1886070014), [2](https://github.com/xenova/transformers.js/issues/338#issuecomment-1742713301)) and [pages router](https://github.com/xenova/transformers.js/issues/338#issuecomment-1743921037).

["pnpm is a bad replacement"](https://github.com/xenova/transformers.js/issues/875#issuecomment-2260597579)

CahidArda committed:
fix: make llamaindex dev & peer dependency

CahidArda committed:
fix: pin llamaindex to 0.5.24

CahidArda committed:
fix: rm llamaindex

CahidArda committed:
feat: add CF worker deploy test

CahidArda committed:
feat: add deployed tests for nexjts vercel-ai-sdk

CahidArda commented:
let's squash merge

CahidArda committed:
fix: fix CF worker deployed env vars

also bumped langchain/community to fix the ratelimit warning

also refreshed the bun lock file

CahidArda committed:
fix: add ai to dependencies

when used locally, the project would work because it can use 'ai' installed in parent dirs. It caused an issue when deployed though.

CahidArda committed:
fix: add env vars for CF workers deployment

ytkimirti committed:
chore: remove forgotten build output in nextjs folder

ytkimirti committed:
chore: ignore some eslint rules for now

mrsamirr commented:
well @CahidArda,
| rag-chat version | llamaindex version | Vercel (npm) |
|------------------|--------------------|--------------|
| 1.6.4            | 0.5.24             | ❌ [missing tiktoken_bg.wasm](https://github.com/run-llama/create-llama/issues/164) |

The issue with `missing tiktoken_bg.wasm` on Vercel has been resolved as mentioned in [this comment](https://github.com/run-llama/create-llama/issues/164#issuecomment-2387573668). After resolving this issue, the rag-chat application now works seamlessly on Vercel.

CahidArda commented:
Hey @mrsamirr,

we can reconsider adding llama parser to rag-chat in the future, but we didn't want to make installing rag-chat require that much configuration at this time.

Thanks a lot on reporting your solution!
